File: packages/expo-app-metrics/ios/LogEvents/DisplayNameValidation.swift
// Copyright 2025-present 650 Industries. All rights reserved.

/// Max length of a log event display name in characters; overlong values are truncated, not dropped.
private let maxDisplayNameLength = 128

/// Validates and normalizes a caller-provided log event display name.
///
/// Trims surrounding whitespace and returns `nil` for `nil` or blank input so the call site
/// can omit the field entirely. Overlong values are truncated with a warning, preserving the prefix.
func validateDisplayName(_ displayName: String?) -> String? {
  guard let displayName else {
    return nil
  }
  let trimmed = displayName.trimmingCharacters(in: .whitespacesAndNewlines)
  if trimmed.isEmpty {
    return nil
  }
  return truncateToMaxLength(
    trimmed,
    maxLength: maxDisplayNameLength,
    warningMessage:
      "[AppMetrics] logEvent truncated displayName from \(trimmed.count) characters "
      + "to the \(maxDisplayNameLength)-character limit."
  )
}

File: packages/expo-app-metrics/android/src/main/java/expo/modules/appmetrics/logevents/DisplayNameValidation.kt
package expo.modules.appmetrics.logevents

// Max length of a log event display name in characters; overlong values are truncated, not dropped.
private const val MAX_DISPLAY_NAME_LENGTH = 128

/**
 * Validates and normalizes a caller-provided log event display name.
 *
 * Trims surrounding whitespace and returns `null` for `null` or blank input so the call site
 * can omit the field entirely. Overlong values are truncated with a warning, preserving the prefix.
 */
internal fun validateDisplayName(displayName: String?): String? {
  if (displayName == null) {
    return null
  }
  val trimmed = displayName.trim()
  if (trimmed.isEmpty()) {
    return null
  }
  return truncateToMaxLength(
    trimmed,
    MAX_DISPLAY_NAME_LENGTH,
    "[AppMetrics] logEvent truncated displayName from ${trimmed.length} characters to the $MAX_DISPLAY_NAME_LENGTH-character limit."
  )
}
